crates/wast/src/resolve/component_names.rs

```diff
@@ -0,0 +1,817 @@
+            }
+            ComponentTypeField::Import(import) => match &import.type_ {
+                ComponentTypeUse::Inline(inline) => match inline {
+                    DefType::Func(f) => {
+                        resolver.funcs.register(f.id, "component func")?;
+                    }
+                    DefType::Module(m) => {
+                        resolver.modules.register(m.id, "module")?;
+                    }
+                    DefType::Component(c) => {
+                        resolver.components.register(c.id, "component")?;
+                    }
+                    DefType::Instance(i) => {
+                        resolver.instances.register(i.id, "instance")?;
+                    }
+                    DefType::Value(v) => {
+                        resolver.values.register(v.id, "value")?;
+                    }
+                },
+                ComponentTypeUse::Ref(_index) => {}
+            },
+            ComponentTypeField::Export(_export) => {}
+        }
+        i += 1;
+    }
+
+    resolve_stack.pop();
+    Ok(())
+}
+
+fn resolve_instance_type<'a, 'b>(
+    c: &'b mut InstanceType<'a>,
+    resolve_stack: &'b mut Vec<ComponentResolver<'a>>,
+) -> Result<(), Error> {
+    resolve_stack.push(ComponentResolver::new(c.id));
+
+    // Iterate through the fields of the component type. We use an index
+    // instead of an iterator because we'll be inserting aliases
+    // as we go.
+    let mut i = 0;
+    while i < c.fields.len() {
+        // Resolve names within the field.
+        match &mut c.fields[i] {
+            InstanceTypeField::Alias(alias) => {
+                resolve_alias(alias, resolve_stack)?;
+            }
+            InstanceTypeField::Type(ty) => {
+                resolve_type_def(&mut ty.def, resolve_stack)?;
+            }
+            InstanceTypeField::Export(export) => {
+                resolve_deftype_use(&mut export.item, resolve_stack)?;
+            }
+        }
+
+        // Name resolution may have emitted some aliases. Insert them before
+        // the current definition.
+        let resolver = resolve_stack.last_mut().unwrap();
+        let aliases_to_insert = replace(&mut resolver.aliases_to_insert, Vec::new());
+        for alias in aliases_to_insert {
+            c.fields.insert(i, InstanceTypeField::Alias(alias));
+            i += 1;
+        }
+
+        // Definitions can't refer to themselves or to definitions that appear
+        // later in the format. Now that we're done resolving this field,
+        // assign it an index for later defintions to refer to.
+        match &mut c.fields[i] {
+            InstanceTypeField::Alias(alias) => {
+                register_alias(alias, resolver)?;
+            }
+            InstanceTypeField::Type(ty) => {
+                resolver.types.register(ty.id, "type")?;
+            }
+            InstanceTypeField::Export(_export) => {}
+        }
+        i += 1;
+    }
+
+    resolve_stack.pop();
```

Comment:
This pattern of push, loop, add aliases, etc shows up in a few places, could this be wrapped up into a helper method to avoid duplicating the logic to insert aliases based on the results of resolution?

Reply:
This is split out to https://github.com/bytecodealliance/wasm-tools/issues/599